File: android/app/src/main/kotlin/fr/fabrique/social/gouv/pass_emploi_app/CvmRepository.kt
package fr.fabrique.social.gouv.pass_emploi_app

import android.content.Context
import androidx.lifecycle.LifecycleOwner
import com.poleemploi.matrixlib.Manager.MatrixManager
import com.poleemploi.matrixlib.Manager.SessionManager
import com.poleemploi.matrixlib.Model.Event
import com.poleemploi.matrixlib.Model.EventType
import com.poleemploi.matrixlib.Model.Room

class CvmRepository(
    private val context: Context,
    private val lifecycleOwner: LifecycleOwner,
) {

    private var room: Room? = null
    private var onMessage: (List<Map<String, Any?>>) -> Unit = {}
    private var onRoom: (Boolean) -> Unit = {}

    fun setEventCallback(callback: (List<Map<String, Any?>>) -> Unit) {
        this.onMessage = callback
    }

    fun setRoomsCallback(callback: (Boolean) -> Unit) {
        this.onRoom = callback
    }

    fun initCvm() {
        MatrixManager.initContext(context)
    }

    fun login(ex160: String, token: String, onResult: (Boolean) -> Unit) {
        try {
            MatrixManager.getInstance().loginAndStartSession(token, ex160)
            onResult(true)
        } catch (ignored: Exception) {
            onResult(false)
        }
    }

    fun joinFirstRoom(onResult: (Boolean) -> Unit) {
        MatrixManager.getInstance().joinFirstRoom(lifecycleOwner) {
            this.room = it
            onResult(true)
        }
    }

    suspend fun sendMessage(message: String, callback: (Boolean) -> Unit) {
        this.room?.let { room ->
            MatrixManager.getInstance().sendMessage(
                room.id!!,
                message,
                { _: String -> callback(true) },
                { _: Int, _: String -> callback(false) }
            )
        } ?: callback(false)
    }


    fun startListenMessages(callback: (Boolean) -> Unit) {
        if (this.room == null) {
            callback(false)
            return
        }
        MatrixManager.getInstance().startListenMessage(room!!.id!!) { events ->
            val allMessages = events.map { event ->
                mapOf(
                    "id" to event.eventId,
                    "type" to stringify(event.eventType),
                    "isFromUser" to (event.senderId == SessionManager.matrixUserId),
                    "message" to event.message,
                    "fileInfo" to event.url,
                    "date" to event.date?.time,
                    "readByConseiller" to event.readByConseiller(),
                    "readByJeune" to event.readByJeune()
                )
            }
            this.onMessage(allMessages)
        }
        callback(true)
    }

    fun loadMore(limit: Int, callback: (Boolean) -> Unit) {
        if (MatrixManager.getInstance().hasMoreMessages == true) {
            MatrixManager.getInstance().loadMoreMessages(limit)
        }
        callback(true)
    }

    suspend fun markAsRead(eventId: String, callback: (Boolean) -> Unit) {
        this.room?.let { room ->
            MatrixManager.getInstance().markAsRead(room.id!!, eventId)
            callback(true)
        } ?: callback(false)
    }

    fun stopListenMessage() {
        MatrixManager.getInstance().stopListenMessage()
    }

    fun startListenRoom() {
        MatrixManager.getInstance().startRoomListener(lifecycleOwner) { rooms ->
            this.room = rooms.firstOrNull()
            this.onRoom(this.room != null)
        }
    }

    fun stopListenRoom() {
        MatrixManager.getInstance().stopRoomListener(lifecycleOwner)
    }

    fun logout() {
        MatrixManager.getInstance().stopSession()
        this.room = null
    }
}

private fun stringify(eventType: EventType): String {
    return when (eventType) {
        EventType.MESSAGE -> "message"
        EventType.FILE -> "file"
        EventType.IMAGE -> "image"
        EventType.TYPING -> "typing"
        else -> "unknown"
    }
}

private fun Event.readByConseiller(): Boolean {
    println("🚀🚀🚀 CvmRepository.kotlin#message ->  $message")
    println("🚀🚀🚀 CvmRepository.kotlin#readBy ->  $readBy")
    val readByConseiller = readBy?.filterNot { it == SessionManager.matrixUserId }?.isNotEmpty() == true
    println("🚀🚀🚀 CvmRepository.kotlin#readByConseiller ->  $readByConseiller")
    println("🚀🚀🚀 ----------")
    return readByConseiller
}

private fun Event.readByJeune(): Boolean {
    return readBy?.any { it == SessionManager.matrixUserId } == true
}

File: ios/Runner/CVM/CvmRepository.swift
import Foundation
import BenedicteSDK

typealias EventJson = Dictionary<String, Any>
typealias EventsReceived = ([EventJson]) -> Void
typealias RoomsReceived = (Bool) -> Void
typealias SuccessCompletion = (Bool) -> Void
typealias VoidCompletion = () -> Void

class CvmRepository {
    
    private var room: Room?
    private var onMessages: EventsReceived?
    private var onHasRooms: RoomsReceived?
    
    func initializeCvm(limit: Int) {
        MatrixManager.sharedInstance.initialize(paginationPageSize: limit)
    }
    
    func setMessageCallback(_ onMessages: @escaping EventsReceived) {
        self.onMessages = onMessages
    }
    
    func setHasRoomCallback(_ onHasRooms: @escaping RoomsReceived) {
        self.onHasRooms = onHasRooms
    }

    func login(token: String, ex160Url: String, completion: @escaping SuccessCompletion) {
        MatrixManager.sharedInstance.loginAndStartSession(accessToken: token, oauthServer: ex160Url, completion: completion)
    }

    func joinFirstRoom(completion: @escaping SuccessCompletion) {
        MatrixManager.sharedInstance.joinFirstRoom { [weak self] room in
            self?.room = room
            completion(room != nil)
        }
    }
    
    func startListenMessages(completion: @escaping SuccessCompletion) {
        guard let room = self.room else {
            completion(false)
            return
        }

        MatrixManager.sharedInstance.startMessageListener(room: room) { [weak self] events in
            self?.onMessages?(events.map { $0.toJson() })
        }
        completion(true)
    }
    
    func stopListenMessages() {
        if let room = room {
            MatrixManager.sharedInstance.stopMessageListener(room: room)
        }
    }
    
    func sendMessage(_ message: String, completion: @escaping SuccessCompletion) {
        guard let room = self.room else {
            completion(false)
            return
        }
        MatrixManager.sharedInstance.sendMessage(room: room, message: message, completion: completion)
    }
    
    func loadMore(limit: Int, completion: @escaping VoidCompletion) {
        guard let room = self.room else {
            completion()
            return
        }
        
        MatrixManager.sharedInstance.loadMoreMessage(room: room, withPaginationSize: limit, completion: completion)
    }
    
    func markAsRead(_ eventId: String, completion: @escaping SuccessCompletion) {
        guard let room = self.room else {
            completion(false)
            return
        }
        
        MatrixManager.sharedInstance.markAsRead(room.id, eventId: eventId, completion: completion)
    }

    func startListenRoom() {
        MatrixManager.sharedInstance.startRoomListener { [weak self] rooms in
            self?.onHasRooms?(rooms?.first != nil)
        }
    }

    func stopListenRoom() {
        MatrixManager.sharedInstance.stopRoomListener()
    }

    func logout() {
        MatrixManager.sharedInstance.stopSession()
        self.room = nil
    }
}

private extension Event {
    func toJson() -> EventJson {
        return [
            "id": eventId,
            "type": eventType.toString(),
            "isFromUser": senderID == SessionManager.sharedInstance.userId,
            "message": message,
            "date": timestamp,
            "fileInfo": attachmentID,
            "readByConseiller": readByConseiller
        ]
    }

    private var timestamp: Int64? {
        guard let date = date else {
            return nil
        }
        return Int64((date.timeIntervalSince1970) * 1000)
    }
    
    private var readByConseiller: Bool{
        print("🚀🚀🚀 CvmRepository.swift#message -> ", message)
        print("🚀🚀🚀 CvmRepository.swift#readBy -> ", readBy)
        return !readBy.filter({ $0 != SessionManager.sharedInstance.userId }).isEmpty
    }
}

private extension EventType {
    func toString() -> String {
        switch self {
        case .UNKNOWN:
            return "unknown"
        case .MESSAGE:
            return "message"
        case .FILE:
            return "file"
        case .IMAGE:
            return "image"
        case .TYPING:
            return "typing"
        case .READ:
            return "read"
        case .MEMBER:
            return "member"
        }
    }
}
